mantidproject/mantid
# Miscellaneous multiple scattering fixes for PEARL validation

**Description of work.**

Set of changes that came from running some multiple scattering calculations on PEARL datasets supplied by Nick Funnel. This revealed some performance issues and also a couple of bugs. See description on linked issue for details.

**To test:**

Unzip the attached input files + update the input folder in the following script.

Run the script which tests a couple of the changes (change to random seed, changes to workspace naming to avoid overwrite) and also acts as a general sanity check. 

Check that the results for spectrum 20 are the same in both output workspace groups. It uses a low number of scenarios so the answer won't be v accurate but you can see the result for spectrum 20 is unchanged. 

Also check that the names of the workspaces in the output workspace groups from the two calls to CalculateMultipleScattering are distinct.
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

def loadPEARLWSAndRunMS(SpectrumList, OutputName):
    inputWS = Load(Filename=os.path.join(folder,'PEARL00115245.nxs'), SpectrumList=SpectrumList) # Ce02
    # S(Q) has been cropped 1500 to 19995 during PEARL data reduction so do same here
    inputWS = CropWorkspace(inputWS, 1500, 19995)
    inputWS = ConvertUnits(inputWS, Target="Wavelength")

    # 5mm x 5mm beam
    SetBeam(inputWS, Geometry={'Shape': 'Slit', 'Width': 0.5, 'Height': 0.5})
    # 6mm diameter can
    SetSample(InputWorkspace=inputWS,Geometry={'Shape': 'Cylinder', 'Height': 4.0,
                        'Radius': 0.2985, 'Center': [0.,0.,0.]},
              Material={'ChemicalFormula': 'Ce-O2', 'MassDensity': 7.215})
           
    CalculateMultipleScattering(InputWorkspace=inputWS,SofqWorkspace=SOfQ,
           NeutronPathsSingle=100,NeutronPathsMultiple=100, NumberScatterings=3, 
           OutputWorkspace=OutputName, Interpolation="Linear", NumberOfWavelengthPoints=300)

folder = r"C:\Multiple Scattering\PEARL tests"
SOfQ = Load(Filename=os.path.join(folder,'PRL115245-115251_tt70_mods1-9_Q.nxs')) #Ce02
# scale S(Q) so it tends to 1 at large q - rough approach for now
high_q = SOfQ.dataY(0)[-9:]
high_q_avg = sum(high_q)/len(high_q)
SOfQ_y = SOfQ.dataY(0)
SOfQ_y /= high_q_avg

SOfQ_x = SOfQ.dataX(0)
average_q_spacing = (SOfQ_x[-1]-SOfQ_x[0])/(SOfQ_x.size-1)
SOfQ=Rebin(SOfQ,[SOfQ_x[0],average_q_spacing,SOfQ_x[-1]])

loadPEARLWSAndRunMS('', "MuscatResults")
loadPEARLWSAndRunMS(20, "MuscatResultsSp20")
```
[PEARLFiles.zip](https://github.com/mantidproject/mantid/files/6724707/PEARLFiles.zip)

Fixes #31892.





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
